cms/djangoapps/contentstore/docs/decisions/0003-hybrid-approach-for-public-apis.rst

```diff
@@ -0,0 +1,67 @@
+0003: Hybrid approach for public course authoring APIs
+======================================================
+
+Context
+-------
+
+We are planning to offer public APIs via oauth that allow course authors to create
+and update course content via an external application. The plan for this is outlined in this
+`Spec Demo`_.
+
+.. _`Spec Demo`: https://openedx.atlassian.net/wiki/spaces/COMM/pages/3696066564/Spec+Memo+API-Based+Management+of+edX+Course+Blocks+Outlines+and+Settings+MVP.
+
+Rather than restricting a course to only ever be editable either via API or manually,
+we will offer a hybrid solution where both is possible.
+
+A hybrid solution faces two challenges:
+
+- "out-of-date write conflicts": how do we handle a situation where an API user updates course content,
+  yet there is an implicit conflict because a manual user edited the course recently and now the API user's local files are outdated?
+  This creates a danger of overwriting another person's changes.
+- "dirty writes": how do we handle a situation where there are multiple asynchronous import or update operations
+  at once that may be in conflict with each other?
+
+Some considerations that come into play when picking an approach are:
+
+1. Having the system reliably play traffic cop isn't simple, principally because of our reliance on slow-running async tasks.
+2. Even if we ignore the problem of playing traffic cop, embracing hybrid editing means clients need some way to know
+   whether other users have made changes.
+
+We analyzed a range of the most viable options to handle these challenges.
+We mainly rejected two alternatives, although there are many options:
```

Comment:
Unclear how to process this. If there are many options, what's the significance of having rejected these two?

Reply:
How about:
We ruled out any we found to be too risky or hard to implement due to a high amount of complexity. That left us with three alternatives, of which we rejected the following two: